Converted initial_sharding to go

arindamnayak opened the pull request:
Converted initial_sharding to go. Following files are converted.
- initial_sharding.py
- initial_sharding_bytes.py
- initial_sharding_multi.py (work in progress)
- initial_sharding_multi_split_diff.py ( work in progress)

UPDATE: Now, everything is fixed now.

arindamnayak committed:
Rebase test_master with latest Cluster code (#22)

* pick latest fix from cluster_for_test

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Rebase test_master with latest Cluster code (#22)

* pick latest fix from cluster_for_test

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Rebase test_master with latest Cluster code (#22)

* pick latest fix from cluster_for_test

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

ajeetj committed:
Tabletmanager2 test cases in GO using cluster (#23)

* tabletmanager lock unlock table test case

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

ajeetj committed:
Tabletmanager2 test cases in GO using cluster (#23)

* tabletmanager lock unlock table test case

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

ajeetj committed:
Tabletmanager2 test cases in GO using cluster (#23)

* tabletmanager lock unlock table test case

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

arindamnayak committed:
Vtctld Test cases in Go using cluster (#25)

* converted vtctld_test.py to go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Vtctld Test cases in Go using cluster (#25)

* converted vtctld_test.py to go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Vtctld Test cases in Go using cluster (#25)

* converted vtctld_test.py to go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted schema.py testcase (#26)

* migrated one of testcase from schema.py to schema_test.go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted schema.py testcase (#26)

* migrated one of testcase from schema.py to schema_test.go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted schema.py testcase (#26)

* migrated one of testcase from schema.py to schema_test.go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

ajeetj committed:
Tablet Manager test cases in Go using cluster (#27)

* converted tabletmanager test cases to go

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

ajeetj committed:
Tablet Manager test cases in Go using cluster (#27)

* converted tabletmanager test cases to go

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

ajeetj committed:
Tablet Manager test cases in Go using cluster (#27)

* converted tabletmanager test cases to go

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted keyspace test to go (#28)

* ported testcase of keyspace_test.py

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted keyspace test to go (#28)

* ported testcase of keyspace_test.py

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted keyspace test to go (#28)

* ported testcase of keyspace_test.py

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted sharded test from py to go (#29)

* Converted sharded test from py to go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted sharded test from py to go (#29)

* Converted sharded test from py to go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Converted sharded test from py to go (#29)

* Converted sharded test from py to go

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
initial commit for sharding test

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added reconciliation assert

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added insert and stream replication

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added rest assertions including migrateservedtypes

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

saurabh408 committed:
Added testcase for mysqlctl process

Signed-off-by: saurabh <saurabh.408@gmail.com>

saurabh408 committed:
Added testcase for mysqlctl process

Signed-off-by: saurabh <saurabh.408@gmail.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
completed remaining assertions and fixed non-workings

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

saurabh408 committed:
removed unused variables

Signed-off-by: saurabh <saurabh.408@gmail.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

saurabh408 committed:
removed unused variables

Signed-off-by: saurabh <saurabh.408@gmail.com>

arindamnayak committed:
converted to v3 way of resharding

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

saurabh408 committed:
removing comment

Signed-off-by: saurabh <saurabh.408@gmail.com>

saurabh408 committed:
removing comment

Signed-off-by: saurabh <saurabh.408@gmail.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
restructured the testcases

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

saurabh408 committed:
addressed review comments

Signed-off-by: saurabh <saurabh.408@gmail.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

saurabh408 committed:
addressed review comments

Signed-off-by: saurabh <saurabh.408@gmail.com>

arindamnayak committed:
added testcase with bytes as sharding key type

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
removed cleandir flag from vttablet teardown

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
removed cleandir flag from vttablet teardown

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added v3 sharding test

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

saurabh408 committed:
vtgate buffer testcase migrated in go

Signed-off-by: saurabh <saurabh.408@gmail.com>

saurabh408 committed:
vtgate buffer testcase migrated in go

Signed-off-by: saurabh <saurabh.408@gmail.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
used rbr mysql mode

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
changed code to use v3 way of sharding

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -70,17 +78,40 @@ type Shard struct {
+}
+
+// Replica get the last but one tablet which is replica
+// Mostly we have either 2 or 3 tablet setup [master, replica], [master, replica, rdonly]
```
this won't work in a 2 tablet setup?

arindamnayak replied:
I will change the comment, as mostly we test with 3 tablets or more.

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -312,3 +389,41 @@ func (cluster *LocalProcessCluster) GetAndReserveTabletUID() int {
 	return int(rand.Int31n(maxNumber)) + baseNumber
 }
+
+// GetVttabletInstance create a new vttablet object
```
create -> creates

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -312,3 +389,41 @@ func (cluster *LocalProcessCluster) GetAndReserveTabletUID() int {
+	}
+}
+
+// StartVttablet start a new tablet
```
start -> starts

deepthi commented on go/test/endtoend/cluster/vtctld_process.go:
```diff
@@ -132,8 +134,8 @@ func (vtctld *VtctldProcess) TearDown() error {
-	os.RemoveAll(path.Join(vtctld.Directory, "tmp"))
-	os.RemoveAll(path.Join(vtctld.Directory, "backups"))
+	os.RemoveAll(vtctld.LogDir)
+	//os.RemoveAll(path.Join(vtctld.Directory, "backups"))
```
delete this line.

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -56,6 +61,10 @@ type VttabletProcess struct {
+	EnableSemiSync              bool
+	SupportBackup               bool
+	ServingStatus               string
+	DbPwd                       string
```
rename to DbPassword

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -82,14 +91,20 @@ func (vttablet *VttabletProcess) Setup() (err error) {
 		"-vtctld_addr", vttablet.VtctldAddress,
 	)
+
+	if vttablet.SupportBackup {
```
rename SupportBackup -> SupportsBackup

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -109,52 +124,109 @@ func (vttablet *VttabletProcess) Setup() (err error) {
 
-	return fmt.Errorf("process '%s' timed out after 60s (err: %s)", vttablet.Name, <-vttablet.exit)
+// GetStatus function checks if vttablet process is up and running
+func (vttablet *VttabletProcess) GetStatus() string {
```
What is the difference between this and `GetTabletStatus`? Why do we need both?

arindamnayak replied:
`GetTabletStatus` gets the tablet status, but `GetStatus` hits `/debug/status` gets html page, where we have some assertion regarding the binlog player.

deepthi replied:
Can you change the comment describing these two functions?

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -168,19 +240,46 @@ func (vttablet *VttabletProcess) TearDown() error {
 	}
 }
 
+// CreateDB created vt_keyspace database
```
`CreateDB creates the database for keyspace`

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -168,19 +240,46 @@ func (vttablet *VttabletProcess) TearDown() error {
+	return err
+}
+
+// QueryTablet lets you execute query in this tablet and get the result
```
lets you execute -> executes a ...

deepthi commented on go/test/endtoend/cluster/vtworker_process.go:
```diff
@@ -0,0 +1,196 @@
+	return false
+}
+
+// TearDown shutdowns the running vtworker service
```
service -> process

deepthi commented on go/test/endtoend/cluster/vtworker_process.go:
```diff
@@ -0,0 +1,196 @@
+	}
+}
+
+// ExecuteCommand executes any vtctlclient command
```
vtctlclient -> vtworker

deepthi commented on go/test/endtoend/cluster/vtworker_process.go:
```diff
@@ -0,0 +1,196 @@
+	return tmpProcess.Run()
+}
+
+// ExecuteVtworkerCommand executes any vtctlclient command
```
vtctlclient -> vtworker

deepthi commented on go/test/endtoend/cluster/vtworker_process.go:
```diff
@@ -0,0 +1,196 @@
+
+// VtworkerProcessInstance returns a vtworker handle
+// configured with the given Config.
+// The process must be manually started by calling setup()
```
Setup()

deepthi commented on go/test/endtoend/sharded/shared_keyspace_test.go:
```diff
@@ -0,0 +1,250 @@
+/*
```
Is the file supposed to be called sharded_keyspace_test.go?

arindamnayak replied:
Done

deepthi commented on go/test/endtoend/sharding/base_sharding.go:
```diff
@@ -0,0 +1,353 @@
+	lotRange2 uint64 = 0xE000000000000000
+	// InsertTabletTemplateKsID common insert format to be used for different tests
+	//
+	InsertTabletTemplateKsID = `insert into %s (id, msg) values (%d, '%s') /* vtgate:: keyspace_id:%016X */ /* id:%d */`
```
Is this still being used? If yes, remove the comments inside the query, those are v2.

arindamnayak replied:
Since we are inserting to tablets directly, we need this - [link to discussion](https://planetscale.slack.com/archives/GNKDZJZ7T/p1575186830032900).

deepthi replied:
no, you don't need these comments. they are only parsed and used while inserting via vtgate, and even there with v3 you should not need them.

deepthi commented on go/test/endtoend/sharding/initialsharding/sharding_util.go:
```diff
@@ -0,0 +1,676 @@
+}
+
+// TestInitialShardingWithVersion - main test with accepts different params for various test
+func TestInitialShardingWithVersion(t *testing.T, keyspace *cluster.Keyspace, shardingKeyType topodata.KeyspaceIdType, isMulti bool, isExternal bool) {
```
we are no longer doing v2, so rename this to just `TestInitialSharding`?
Instead of using KeyspaceIdType (uint64 or bytes), maybe we should be using the type of the column directly here. KeyspaceIdType is a v2 concept.

arindamnayak replied:
To accommodate `bytes[]` and reuse existing test, I passed this as parameter.

deepthi replied:
I understand. My point is:
v2 tests use ShardingColumnType, so this had to be set to UINT64 or BYTES. v3 tests don't need to use that specific type. The distinction in v3 is the column type in the tablet itself - whether it is int or varbinary. We can use some other flag to denote that, it should not be tied to ShardingColumnType. Either it can be a string ("integer" vs "string"), or it can be https://github.com/vitessio/vitess/blob/master/go/sqltypes/type.go#L134

arindamnayak replied:
Ok, I will do that. I was using that as flag, I have not used it to set any sharding column.

deepthi reviewed (commented):
Looking good so far.

arindamnayak committed:
added multi split diff and multi sharding

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

deepthi commented on go/test/endtoend/cluster/vtgate_process.go:
```diff
@@ -80,7 +80,7 @@ func (vtgate *VtgateProcess) Setup() (err error) {
 		"-service_map", vtgate.ServiceMap,
 		"-mysql_auth_server_impl", vtgate.MySQLAuthServerImpl,
-		"-pid_file", vtgate.PidFile,
+		//"-pid_file", vtgate.PidFile,
```
what's the problem with the pid_file?

arindamnayak replied:
I did this during some trial to fix the last 2 testcases (multisplitdiff, multi). I will revert once all things are settled.

ajeetj committed:
TabletManager remaining test cases

TabletManager remaining test cases in GO
Signed-off-by: Ajeet jain <ajeet@planetscale.com>

ajeetj committed:
TabletManager remaining test cases

TabletManager remaining test cases in GO
Signed-off-by: Ajeet jain <ajeet@planetscale.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
moved sql start to non-blocking mode (#35)

* moved sql start to non-blocking mode

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
moved sql start to non-blocking mode (#35)

* moved sql start to non-blocking mode

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added testcase for vtgate vars

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added testcase for vtgate vars

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Merge branch 'tal_test_master' into tal_test_initial_sharding_v3

arindamnayak committed:
implement pr suggestion

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
use ks1 mysql ports for other tablets in ks2

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
redirect process errors to log files (#38)

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
debug the test error in ci

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
fix initial sharding multi

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak commented:
@deepthi , all pr suggestions are implemented, those are listed below.
- renamed shared_keyspace_test to sharded_keyspace_test
- remove `/*vtgate ... */` from insert query
- replace usage of shardingkeytype to `querypd.type`
- Changed some of comments

Now, initiali_sharding_multi is working fine. Only multisplitdiff is remaining.

arindamnayak committed:
Merge branch 'tal_test_master' into tal_test_initial_sharding_v3

arindamnayak committed:
fixed protected mysqlctl teardown

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
refactor mysqlctl start and stop

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Fixed multisplitdiff command

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added log statement for error condition

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

ajeetj committed:
Ignoring Python tests which are moved to GO (#40)

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

* readme for go endtoend test cases

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

* Update README.md

arindamnayak committed:
Merge branch 'tal_test_master' into tal_test_initial_sharding_v3

arindamnayak committed:
updated comments

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Refactor code

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak commented:
After I rebased the base branch of this PR (Tal_test_master) with vitessio/master, this PR seems to be inconsistent, i.e. it is reporting some new files are added, whereas I added files only related to initial sharding. So please consider the initial_sharding related change for review. Also there is conflicting files, which I will fix once the parent PR is merged to upstream.

arindamnayak committed:
Merge branch 'tal_test_master' into tal_test_initial_sharding_v3

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added more error checks

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

ajeetj commented:
LGTM
@deepthi I am merging this in tal_test_master so that we can create PR for Sugu in vitess.io
